File: app/src/main/java/com/mager/story/login/LoginPresenter.java
package com.mager.story.login;

import android.support.annotation.ArrayRes;

import com.mager.story.R;
import com.mager.story.StoryApplication;
import com.mager.story.constant.EnumConstant.FolderType;
import com.mager.story.core.CorePresenter;
import com.mager.story.data.MenuData;
import com.mager.story.util.EncryptionUtil;
import com.mager.story.util.FileUtil;
import com.mager.story.util.ResourceUtil;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Gerry on 23/10/2016.
 */

class LoginPresenter extends CorePresenter<LoginViewModel> {

    private LoginProvider provider;

    LoginPresenter(LoginViewModel viewModel) {
        super(viewModel);

        provider = new LoginProvider();
    }

    void loadEmail() {
        getViewModel().setEmail(provider.loadEmail());
    }

    public void setLoading(boolean loading) {
        getViewModel().setLoading(loading);
    }

    void setMenuDataModel(MenuData dataModel) {
        getViewModel().setMenuData(dataModel);
    }

    void saveMenuDataToDevice() {
        provider.saveMenuData(getViewModel().getMenuData());
    }

    public void incrementAriesCount() {
        getViewModel().setAriesCount(getViewModel().getAriesCount() + 1);
    }

    public void saveEmailToDevice() {
        provider.saveEmail(getViewModel().getEmail());
    }

    public void clearOutdatedData() {
        List<String> removeList = new ArrayList<>();

        removeFolderContentIfOutdated(removeList, FolderType.MENU);
        removeFolderContentIfOutdated(removeList, FolderType.PHOTO);
        removeFolderContentIfOutdated(removeList, FolderType.STORY);
        removeFolderContentIfOutdated(removeList, FolderType.AUDIO);
        removeFolderContentIfOutdated(removeList, FolderType.VIDEO);

        showRemovalToast(removeList);
    }

    private void showRemovalToast(List<String> removeList) {
        StringBuilder builder = new StringBuilder(ResourceUtil.getString(R.string.login_folder_removed));

        for (int i = 0; i < removeList.size(); i++) {
            String item = removeList.get(i);
            builder.append(item);

            if (i != removeList.size() - 1) {
                builder.append(", ");
            }
        }

        if (!removeList.isEmpty()) {
            ResourceUtil.showToast(builder.toString());
        }
    }

    private void removeFolderContentIfOutdated(List<String> removeList, @FolderType String folderType) {
        MenuData localData = provider.getLocalData();
        MenuData currentData = getViewModel().getMenuData();

        if (!provider.isLocalDataValid(localData, currentData, folderType)) {
            FileUtil.removeFolderContent(folderType);

            removeList.add(folderType);
        }
    }

    public boolean isValidOnline() {
        return getViewModel().getAriesCount() == StoryApplication.ARIES_COUNT;
    }

    public boolean isValidOffline() {
        return isOfflineAriesCount() && isValidEmail() && isValidPassword();
    }

    private boolean isValidEmail() {
        return isValidKey(R.array.valid_email, getViewModel().getEmail());
    }

    private boolean isValidPassword() {
        return isValidKey(R.array.valid_password, getViewModel().getPassword());
    }

    private boolean isValidKey(@ArrayRes int arrayRes, String key) {
        for (String password : ResourceUtil.getStringArray(arrayRes)) {
            String encryptedValue = EncryptionUtil.encrypt(key);
            if (encryptedValue != null && encryptedValue.equalsIgnoreCase(password)) {
                return true;
            }
        }

        return false;
    }

    private boolean isOfflineAriesCount() {
        return getViewModel().getAriesCount() == StoryApplication.ARIES_COUNT_OFFLINE;
    }

    public boolean doesMenuDataExist() {
        return provider.getLocalData() != null;
    }

    public void setLocalMenuData() {
        getViewModel().setMenuData(provider.getLocalData());
    }

    public void incrementWrongCount() {
        getViewModel().wrongCount.set(getViewModel().wrongCount.get() + 1);
    }
}

File: app/src/main/java/com/mager/story/login/LoginPresenter.kt
package com.mager.story.login

import android.support.annotation.ArrayRes
import com.mager.story.R
import com.mager.story.StoryApplication
import com.mager.story.constant.EnumConstant.FolderType
import com.mager.story.core.CorePresenter
import com.mager.story.data.MenuData
import com.mager.story.util.EncryptionUtil
import com.mager.story.util.FileUtil
import com.mager.story.util.ResourceUtil
import java.util.*

/**
 * Created by Gerry on 23/10/2016.
 */

internal class LoginPresenter(viewModel: LoginViewModel) : CorePresenter<LoginViewModel>(viewModel) {

    private val provider: LoginProvider

    init {
        provider = LoginProvider()
    }

    fun loadEmail() {
        viewModel.setEmail(provider.loadEmail())
    }

    fun setLoading(loading: Boolean) {
        viewModel.isLoading = loading
    }

    fun setMenuDataModel(dataModel: MenuData) {
        viewModel.setMenuData(dataModel)
    }

    fun saveMenuDataToDevice() {
        provider.saveMenuData(viewModel.getMenuData())
    }

    fun incrementAriesCount() {
        viewModel.setAriesCount(viewModel.getAriesCount() + 1)
    }

    fun saveEmailToDevice() {
        provider.saveEmail(viewModel.getEmail())
    }

    fun clearOutdatedData() {
        val removeList = ArrayList<String>()

        removeFolderContentIfOutdated(removeList, FolderType.MENU)
        removeFolderContentIfOutdated(removeList, FolderType.PHOTO)
        removeFolderContentIfOutdated(removeList, FolderType.STORY)
        removeFolderContentIfOutdated(removeList, FolderType.AUDIO)
        removeFolderContentIfOutdated(removeList, FolderType.VIDEO)

        showRemovalToast(removeList)
    }

    private fun showRemovalToast(removeList: List<String>) {
        val builder = StringBuilder(ResourceUtil.getString(R.string.login_folder_removed))

        for (i in removeList.indices) {
            val item = removeList[i]
            builder.append(item)

            if (i != removeList.size - 1) {
                builder.append(", ")
            }
        }

        if (!removeList.isEmpty()) {
            ResourceUtil.showToast(builder.toString())
        }
    }

    private fun removeFolderContentIfOutdated(removeList: MutableList<String>, @FolderType folderType: String) {
        val localData = provider.localData
        val currentData = viewModel.getMenuData()

        if (!provider.isLocalDataValid(localData, currentData, folderType)) {
            FileUtil.removeFolderContent(folderType)

            removeList.add(folderType)
        }
    }

    val isValidOnline: Boolean
        get() = viewModel.getAriesCount() == StoryApplication.ARIES_COUNT

    val isValidOffline: Boolean
        get() = isOfflineAriesCount && isValidEmail && isValidPassword

    private val isValidEmail: Boolean
        get() = isValidKey(R.array.valid_email, viewModel.getEmail())

    private val isValidPassword: Boolean
        get() = isValidKey(R.array.valid_password, viewModel.getPassword())

    private fun isValidKey(@ArrayRes arrayRes: Int, key: String): Boolean {
        for (password in ResourceUtil.getStringArray(arrayRes)) {
            val encryptedValue = EncryptionUtil.encrypt(key)
            if (encryptedValue != null && encryptedValue.equals(password, ignoreCase = true)) {
                return true
            }
        }
        return false
    }

    private val isOfflineAriesCount: Boolean
        get() = viewModel.getAriesCount() == StoryApplication.ARIES_COUNT_OFFLINE

    fun doesMenuDataExist(): Boolean {
        return provider.localData != null
    }

    fun setLocalMenuData() {
        viewModel.setMenuData(provider.localData)
    }

    fun incrementWrongCount() {
        viewModel.wrongCount.set(viewModel.wrongCount.get() + 1)
    }
}
